### PR TITLE
Add quota to agent created datasets

### PR DESCRIPTION
**Pull request by leftwo**
This adds a quota and a reservation to crucible datasets created when a region
is created in the crucible agent.  A somewhat arbitrary .25 overhead is reserved
for each region.  17% is the largest possible space the metadata will take, and this
gives us a little room in the unlikely event that all regions are full.

This does not consider snapshot overhead.
 
For issue #605

### CODE REVIEW

**Comment by leftwo** [comment]
After a discussion with @ahl , I've updated the comments a bit, and changed the quota to be larger
than the reservation.

File: agent/src/main.rs [hunk 1]
```diff
@@ -13,6 +13,23 @@ use std::sync::Arc;
...
+ * situations to avoid allocation of a region that we don't have the
+ * space for if that region was to fill up.
+ * As for how much we reserve, it's the region size itself, plus the
+ * amount we need for metadata (currently around 17%) then an additional
+ * 8% that, if things are behaving, we will have for snapshots and
+ * to prevent us from using all the data in the pool.
+ *
+ * In addition, we throw a quota of 3x the region size.  If the region
+ * has grown that big, then something is wrong and we should prevent it
+ * from growing any larger and impacting other regions in the dataset.
```
Comment (ahl) [inline agent/src/main.rs:29]:
```suggestion
 * In addition, we throw a quota of 3x the region size. The failure modes of exhausting the quota are potentially severe and not (as yet) well-tested so we want some buffer between the reservation and the quota. If the region
 * has grown that big however, something is potentially very wrong and we should prevent it
 * from growing any larger and impacting other regions on the disk.
```

File: agent/src/main.rs [hunk 2]
```diff
@@ -13,6 +13,23 @@ use std::sync::Arc;
...
+ */
+const METADATA_BUFFER: f64 = 1.25;
+const REGION_QUOTA: u64 = 3;
```
Comment (ahl) [inline agent/src/main.rs:32]: maybe: `RESERVATION_FACTOR` and `QUOTA_FACTOR`? I leave it to you... just a thought

File: agent/src/main.rs [hunk 3]
```diff
@@ -117,6 +140,63 @@ impl ZFSDataset {
             bail!("zfs create failed! out:{} err:{}", out, err);
         }
 
+        // If requested, set a quota and reservation here.
+        if let Some(reservation) = reservation {
```
Comment (ahl) [inline agent/src/main.rs:144]: I'd suggest we might want something like this:

```rust
        let mut cmd = std::process::Command::new("zfs")
            .arg("create");

    if let Some(reservation) = reservation {
        cmd = cmd.arg("-o").arg(format!("reservation={}", reservation));
    }

        cmd = cmd
            .arg(&dataset)
            .output()?;

```

i.e. I think you can create the dataset with the properties preset.

**Review by ahl — APPROVED** [review]

File: agent/src/main.rs (hunk 2 above)
Comment (leftwo) [reply agent/src/main.rs]: Done

File: agent/src/main.rs (hunk 3 above)
Comment (leftwo) [reply agent/src/main.rs:141]: Yeah, I like it.  Fewer lines of code and easier cleanup.

File: agent/src/main.rs [hunk 4]
```diff
@@ -13,6 +13,23 @@ use std::sync::Arc;
 
 const PROG: &str = "crucible-agent";
 const SERVICE: &str = "oxide/crucible/downstairs";
+/*
+ * As a safety mechanism to prevent the agent from creating more regions
+ * than we have physical space, we using a zfs reservation on each dataset
```
[the remark this answers is not in the document]
Comment (leftwo) [reply agent/src/main.rs]: Done

File: agent/src/main.rs [hunk 5]
```diff
@@ -13,6 +13,23 @@ use std::sync::Arc;
...
+ * we create.  While not a complete solution, it does help in some
+ * situations to avoid allocation of a region that we don't have the
+ * space for if that region was to fill up.
```
[the remark this answers is not in the document]
Comment (leftwo) [reply agent/src/main.rs]: Done

File: agent/src/main.rs (hunk 1 above)
Comment (leftwo) [reply agent/src/main.rs]: Done

**Pull request merged**